examples/fluids/src/dm_utils.c

```diff
@@ -96,6 +47,202 @@ PetscErrorCode DMFieldToDSField(DM dm, DMLabel domain_label, PetscInt dm_field,
   PetscFunctionReturn(PETSC_SUCCESS);
 }
 
+/**
+  @brief Create `CeedElemRestriction` from `DMPlex`.
+
+  Not collective across MPI processes.
+
+  @param[in]   ceed          `Ceed` context
+  @param[in]   dm            `DMPlex` holding mesh
+  @param[in]   domain_label  `DMLabel` for `DMPlex` domain
+  @param[in]   label_value   Stratum value
+  @param[in]   height        Height of `DMPlex` topology
+  @param[in]   dm_field      Index of `DMPlex` field
+  @param[out]  restriction   `CeedElemRestriction` for `DMPlex`
+
+  @return An error code: 0 - success, otherwise - failure
+**/
+PetscErrorCode DMPlexCeedElemRestrictionCreate(Ceed ceed, DM dm, DMLabel domain_label, PetscInt label_value, PetscInt height, PetscInt dm_field,
+                                               CeedElemRestriction *restriction) {
+  PetscInt num_elem, elem_size, num_dof, num_comp, *restriction_offsets_petsc;
+  CeedInt *restriction_offsets_ceed = NULL;
+
+  PetscFunctionBeginUser;
+  PetscCall(
+      DMPlexGetLocalOffsets(dm, domain_label, label_value, height, dm_field, &num_elem, &elem_size, &num_comp, &num_dof, &restriction_offsets_petsc));
+  PetscCall(IntArrayP2C(num_elem * elem_size, &restriction_offsets_petsc, &restriction_offsets_ceed));
+  PetscCallCeed(ceed, CeedElemRestrictionCreate(ceed, num_elem, elem_size, num_comp, 1, num_dof, CEED_MEM_HOST, CEED_COPY_VALUES,
+                                                restriction_offsets_ceed, restriction));
+  PetscCall(PetscFree(restriction_offsets_ceed));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/**
+  @brief Create `CeedElemRestriction` from `DMPlex` domain for mesh coordinates.
+
+  Not collective across MPI processes.
+
+  @param[in]   ceed          `Ceed` context
+  @param[in]   dm            `DMPlex` holding mesh
+  @param[in]   domain_label  Label for `DMPlex` domain
+  @param[in]   label_value   Stratum value
+  @param[in]   height        Height of `DMPlex` topology
+  @param[out]  restriction   `CeedElemRestriction` for mesh
+
+  @return An error code: 0 - success, otherwise - failure
+**/
+PetscErrorCode DMPlexCeedElemRestrictionCoordinateCreate(Ceed ceed, DM dm, DMLabel domain_label, PetscInt label_value, PetscInt height,
+                                                         CeedElemRestriction *restriction) {
+  DM        dm_coord;
+  PetscBool is_simplex = PETSC_FALSE;
+
+  PetscFunctionBeginUser;
+  PetscCall(DMGetCellCoordinateDM(dm, &dm_coord));
+  if (!dm_coord) {
+    PetscCall(DMGetCoordinateDM(dm, &dm_coord));
+  }
+  PetscCall(DMPlexIsSimplex(dm_coord, &is_simplex));
+  if (!is_simplex) PetscCall(DMPlexSetClosurePermutationTensor(dm_coord, PETSC_DETERMINE, NULL));
```

Comment:
Why not call this only once when we create the DM instead of calling every time we make a restriction?

Reply:
We only get the coordinate restriction a couple of times, but this still feels like DM - creating a restriction shouldn't mutate the DM

Reply:
Because the CellCoordinateDM isn't necessarily setup in the `DMSetupByOrderEnd_FEM`. It's only an issue when periodic faces are specified. Maybe we should setup both CoordinateDM and CellCoordinateDM during setup? I'm not familiar with why both exist or why you'd want to choose one over the other (excluding the CoordinateDM with periodic faces is wrong).

Reply:
This still  feels like the wrong place to do mutation. Can we push up the setting up of the CellCoordinateDM to some time before this?

Reply:
You're right, it's definitely not the best place for it. Trying to figure out the best way to integrate CellCoordinateDM into the `DMSetupByOrder{Begin,End}_FEM` right now. Can't tell if it needs to go through the same coordinate projection as the normal CoordinateDM.

Reply:
Alright, I've moved the `DMPlexSetClosurePermutationTensor` for the cell coordinate DM to `DMSetupByOrderEnd_FEM`.